examples/recursive-types-2/F.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | This is similar to the @recursive-types@ example, except that the @Tree@ type
+-- is monomorphic. This makes it slightly more complicated, because in the other
+-- example, we could define the @TargetOb@ instance for @Tree@ as
+--
+--  @
+--     type instance TargetOb (Tree a) = Tree (TargetOb a)
+--  @
+--
+-- while here we have to create another data type, @TreeAux@, and use it to define
+-- @TargetOb Tree@.
+--
+-- Note that the following won't work:
+--
+--  @
+--     -- Won't work because Tree does not have CGeneric instance
+--     type instance TargetOb Tree = TargetOb (CG.Rep Tree ())
+--
+--     -- Won't work because this doesn't terminate (Tree is recursive)
+--     type instance TargetOb Tree = TargetOb (Categorifier.Client.Rep Tree)
+--  @
+--
+-- The generated C code is the same as the @recursive-types@ eaxmple.
+module F (fCategorified) where
+
+import qualified Categorifier.C.CExpr.Cat as C
+import Categorifier.C.CExpr.Cat.TargetOb (TargetOb)
+import Categorifier.C.CTypes.CGeneric (CGeneric)
+import qualified Categorifier.C.CTypes.CGeneric as CG
+import Categorifier.C.CTypes.GArrays (GArrays)
+import Categorifier.C.KTypes.C (C)
+import qualified Categorifier.Categorify as Categorify
+import qualified Categorifier.Client as Client
+import Data.Int (Int32, Int64)
+import GHC.Generics (Generic)
+
+data Input = Input
+  { iInt32 :: C Int32,
+    iInt64 :: C Int64
+  }
+  deriving (Generic)
+
+Client.deriveHasRep ''Input
+
+instance CGeneric Input
+
+instance GArrays C Input
+
+type instance TargetOb Input = TargetOb (CG.Rep Input ())
+
+data Output = Output
+  { oInt64 :: C Int64,
+    oBool :: Bool
+  }
+  deriving (Generic)
+
+Client.deriveHasRep ''Output
+
+instance CGeneric Output
+
+instance GArrays C Output
+
+type instance TargetOb Output = TargetOb (CG.Rep Output ())
+
+data Tree = Empty | Branch (C Int32) Tree Tree
+  deriving (Generic)
+
+Client.deriveHasRep ''Tree
+
+data TreeAux = EmptyAux | BranchAux (TargetOb Int32) TreeAux TreeAux
+  deriving (Generic)
+
+Client.deriveHasRep ''TreeAux
+
+type instance TargetOb Tree = TreeAux
+
+f :: Input -> Output
+f = h . g
+
+g :: Input -> Tree
+g inp = Branch (iInt32 inp) (Branch (fromIntegral (iInt64 inp)) Empty Empty) Empty
+
+h :: Tree -> Output
+h = \case
+  Empty -> Output {oInt64 = 0, oBool = False}
+  Branch x l r ->
+    Output
+      { oInt64 = fromIntegral x + oInt64 (h l) + oInt64 (h r),
+        oBool = True
+      }
+
+fCategorified :: Input `C.Cat` Output
+fCategorified = Categorify.expression f
```

Comment:
Why not `Categorify.function` here (or, even better, `Categorify.separately`)?

Reply:
> Why not Categorify.function here

Hmm I wasn't even aware of the existence of `Categorify.function`. All examples currently use `expression`, and I can change them together in a separate commit.

> or, even better, Categorify.separately

This example does not involve categorizing a subpart of a function separately, so I'm not sure why `separately` is a good fit

Reply:
I think (hope) it's mentioned in the docs somewhere, but `Categorify.separately` is what we recommend in general (and is used throughout KH).

`Categorify.function 'f [t|C.Cat|] []` basically produces what you wrote, but you don't have to specify the input and output types, which makes it more robust to changes. `Categorify.separately` does that then goes a step further by providing a `NativeCat` instance, so that anything that gets added down the road that calls `f` won’t re-categorify `f`. So, they're both just future-proofing features over `Categorify.expression`.

For examples, it can a bit tricky, because you want to balance clarity of the example with writing code in a way that scales to larger projects. So even though we don't do any sub-part stuff here, it seems worth writing it in the way you would write it in a project. And maybe it's worth making some changes in `Categorify`, like don’t export the existing `function`, then rename `separately` to `function`. So the `NativeCat` behavior is just a free benefit. I can't think of why one wouldn't want it … right?

Reply:
But also, yeah, I'm happy with that being a later commit.

Reply:
> but Categorify.separately is what we recommend in general (and is used throughout KH).

Yeah that's why I wasn't aware of `function` :smile: